### PR TITLE
Add in proper command line args and a standard configuration file format

### LINKED ISSUE #2: Command-line args and settings
In order to make this more scriptable, I was thinking to add command-line args in order to be able to set all of the settings so that you could do something like this:

```
control_master.py --hostname localhost --portnum 12021 ... --run-all
```

Related to that, using the configparser module to parse config files instead of writing out a python file seems like the better way of parsing.

Would this be a reasonable change to make?

### PR DESCRIPTION
**Pull request by rm5248**
Added in a proper command line help for configuring and running entirely from the command line.

A config file may also be specified instead of using the command line.  Any command-line arguments take precedence over what is defined in the config file.

I'm not a python expert, so I'm sure some parts of what I've written are somewhat strange.

Fixes #2

### CODE REVIEW

**Comment by bobjacobsen** [comment]
@rm5248 Thank you!  Am traveling for a couple days, but will test this and merge ASAP.

**Comment by bobjacobsen** [comment]
@rm5248  I'm having trouble getting this to work.  I'm using Python 3.10 on a Mac.

1) Command line options give errors:

```
% python3.10 control_master.py --device /dev/cu.usbserial-D30A45AZ --run-all --target-node-id 02.01.2c.01.09.00
Traceback (most recent call last):
  File "/Users/jake/Documents/Trains/JMRI/projects/openlcb/OlcbChecker/control_master.py", line 243, in <module>
    main()
  File "/Users/jake/Documents/Trains/JMRI/projects/openlcb/OlcbChecker/control_master.py", line 138, in main
    configure.global_config.devicename = args.device
AttributeError: can't set attribute 'devicename'
```

2) I can't get it to work with a configuration file any more:

With this file:
```
% cat localoverrides.py 
hostname = None
portnumber = 12021
devicename = "/dev/cu.usbserial-D30A45AZ"
targetnodeid = "02.01.2C.01.09.00"
ownnodeid = "03.00.00.00.00.01"
checkpip = True
skip_interactive = "True"
trace = 10
```

it fails to open the port:

```
% python3.10 control_master.py 

OpenLCB checking program
 s Setup

 0 Frame Transport checking
 1 Message Network checking
 2 SNIP checking
 3 Event Transport checking
 4 Datagram Transport checking
 5 Memory Configuration checking
 6 CDI checking
  
 a run all in sequence
  
 q  Quit
>> 0
Traceback (most recent call last):
  File "/Users/jake/Documents/Trains/JMRI/projects/openlcb/OlcbChecker/control_master.py", line 243, in <module>
    main()
  File "/Users/jake/Documents/Trains/JMRI/projects/openlcb/OlcbChecker/control_master.py", line 118, in main
    run_prompt()
  File "/Users/jake/Documents/Trains/JMRI/projects/openlcb/OlcbChecker/control_master.py", line 188, in run_prompt
    import control_frame
  File "/Users/jake/Documents/Trains/JMRI/projects/openlcb/OlcbChecker/control_frame.py", line 7, in <module>
    import check_fr10_init
  File "/Users/jake/Documents/Trains/JMRI/projects/openlcb/OlcbChecker/check_fr10_init.py", line 17, in <module>
    import olcbchecker.setup
  File "/Users/jake/Documents/Trains/JMRI/projects/openlcb/OlcbChecker/olcbchecker/setup.py", line 33, in <module>
    s.connect(configure.global_config.devicename)
  File "/Users/jake/Documents/Trains/JMRI/projects/openlcb/OlcbChecker/openlcb/canbus/seriallink.py", line 24, in connect
    self.port.reset_input_buffer()  # drop anything that's just sitting there already  # noqa: E501
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/serial/serialposix.py", line 682, in reset_input_buffer
    raise PortNotOpenError()
serial.serialutil.PortNotOpenError: Attempting to use a port that is not open
```

The problem here seems to be that the device name is null when it gets to the actual open call.  

Could you take a look at these two invocations please?

**Comment by rm5248** [comment]
Whoops, the devicename should be fixed now.

The `localoverrides.py` method shouldn't work anymore, it's designed to use a config file with the `configparser` python module.  The following file should be equivalent:

```
[OlcbChecker]
devicename=/dev/cu.usbserial-D30A45AZ
targetnodeid=02.01.2C.01.09.00
ownnodeid=03.00.00.00.00.01
checkpip=True
skip_interactive=True
```

See the sample config files for more info.  Pass the file in with the `--config-file` parameter.  The intention with the config files is that you can now have several pre-set configurations that you can run just by changing the config file that you're using(CLI arguments will override parameters given in the config file)

**Comment by rm5248** [comment]
Do you still have concerns Bob?

**Comment by bobjacobsen** [comment]
Did you see the PR I posted on your copy of this?

https://github.com/rm5248/OlcbChecker/pull/1

**Comment by rm5248** [comment]
I did not, I did not get a notification for that PR.  I will take a look at it shortly.

**Comment by rm5248** [comment]
Merged that in.  You should be able to push to the branch this PR if from though, since by default github will let maintainers push to PR branches.

To answer your questions though:

> When I try to run via a network connection to a CS-105, I get an error:

That was a problem with the `--port` argument, fixed now.

> 2. Running with a LocoBuffer was sort-of working. I can get it to run through on some hardware with --run-all set. [...]
> 3. Also, I don't seem to be able to run individual test files [...]

Currently you can get to the menu if you don't pass any command line arguments and it should still work.  Personally I'm not a big fan of menus like that, so it's possible to select the tests via the command line.  For example, you can select what tests to run like the following:

```
./control_message.py  --skip-interactive --hostname localhost --port 12021 --add-test control_frame --add-test control_message
```

This is somewhat limited at the moment, since it is all of the tests for a certain class of tests.  But it could be easily extended by doing something like `--add-test control_frame.fr20_ame` if testing just a single thing makes sense.

Selecting it from the command line or from the config file is designed to make it scriptable.

**Comment by bobjacobsen** [comment]
I'm sorry but I think I've lost track of what we're trying to do here.

Is this about getting rid of the menus?  Or it is about being able to run more than one test from a single command?

There already are some command line options for the setup:

```
% ./control_master.py -h

Available options are:

-h, --help print this message

-a, --address followed by a host[:port] IP address for GridConnect access.
                This is mutually exclusive with the -d --device option
-d, --device followed by a serial port device name.
                This is mutually exclusive with the -h --hostname option
-t, --targetnode followed by a nodeID for the device being checked in 01.23.45.67.89.0A form

-r, --run run the tests instead of presenting a menu

Less frequently needed:

-T, --trace followed by an integer trace level.  Higher numbers are more informatiom.
                0 is just error messages, 10 includes success messages, 20 includes message traces, 
                30 includes frame traces, 40 includes physical layer traces, 50 includes internal code traces
-o, --ownnode followed by a nodeID for the checking device in 01.23.45.67.89.0A form
-p do not check results against PIP bits
-P do check results against PIP bits
-I execute interactive tests
-i skip interactive tests
```

The `-T 0` doesn't work right, as non-error messages are still be emitted.

So this works to run the tests without using the menu:
```
% ./control_frame.py -a 192.168.16.212 -t 09.00.99.03.00.35 -r
% ./control_message.py -a 192.168.16.212 -t 09.00.99.03.00.35 -r
```
but they're a little verbose, e.g.

```
% ./control_message.py -a 192.168.16.212 -t 09.00.99.03.00.35 -r

Node Initialized checking
Interactive test skipped

Verify Node checking
Passed

Protocol Support Inquiry checking
PIP reports:
  Datagram Protocol
  Memory Configuration Protocol
  Event Exchange Protocol
  Adcdi Protocol
  Simple Node Identification Protocol
  Configuration Description Information
Passed

Optional Interaction Rejected checking
Failure - Did not receive Optional Interaction Rejected reply

Duplicate Node ID Discovery checking
Interactive test skipped

At least one message check failed
```

And you need to use multiple commands if you want to run some specific selection of tests.

I'd like to keep the menus working as an option, because that's the way the existing NMRA DCC checker programs run, and those people are familiar with doing that on Windows:

```
% ./control_message.py -a 192.168.16.212 -t 09.00.99.03.00.35   

Message Network Standard checking
 a Run all in sequence
 1 Node Initialized checking
 2 Verify Node checking
 3 Protocol Support Inquiry checking
 4 Optional Interaction Rejected checking
 5 Duplicate Node ID Discovery checking
  
 q go back
```

Or am I missing something else?  Sorry for my confusion.

**Comment by rm5248** [comment]
I may have been a little too aggressive with this, but basically what I want to be able to do is the following:

- Exit with a success code or failure code depending on if the test(s) passed or failed.
- Select which tests to run so it is not all-or-nothing

This is so that for my LCC implementation I can run the tests automatically on build and have my unit tests succeed or fail as appropriate.  The tests need to be granular enough that I can debug individual tests and validate that things are working.  Ideally this would be for each individual test as well, but I did it so that it is each "class" of test so far(control_frame, control_message, control_snip....).

Some of the updates are to make the code a bit cleaner(IMO) - these are somewhat optional to make, so they can be removed if needed, but it wasn't easy to do just one thing and not do the rest:

- Importing things at the start.  Fail-fast is better for development in my opinion, since you don't need to interact with the program in order for the import to fail
- Using the Python [argparse](https://docs.python.org/3/library/argparse.html) module to make parsing easier than using getopt.
- Removing the `localoverrides.py` configuration file mechanism in favor of just a normal configuration file using the [configparser](https://docs.python.org/3/library/configparser.html) module.  I don't want to have to modify code to change settings when I could just use a different config file.  Telling people to modify code to change settings like that is just inviting trouble - if this is installed on a Linux system as a utility, you wouldn't be able to modify a `localoverrides.py` file anyway because of a lack of permissions(assuming it is installed under `/usr/share/olcbchecker` and not under the user's home directory)

**Comment by bobjacobsen** [comment]
Thanks for describing that in detail.

I've fixed a bug in the return codes, now in the HEAD of the `main` branch.  I think the code now satisfies your first two points:

> Exit with a success code or failure code depending on if the test(s) passed or failed.

> Select which tests to run so it is not all-or-nothing

The selection can be done at several levels.  You can run a single check with e.g. `./check_me10_init.py -r`, `./check_me20_verify.py -r`, etc (plus any necessary arguments to specify the connection, remote node ID, whether to run interactive tests, etc).  Or you can run a family of checks with e.g. `./control_message.py -r` or run all families with `./control_master.py -r`

I think we should focus on getting this to run for your purposes, and then move on to cleaning it up.   I'm only able to spend time on it intermittently, unfortunately, as there are a couple other things using up a bunch of my hobby time right now.  

Top of my list on this right now is to fix the `control_*.py` checks that are still extra verbose when `-T 0` is specified.  That flag should suppress normal, non-error output.  AFAIK that's done properly in the individual `check_*.py` files, just not the `control_*.py` ones.  But I could certainly work on something else instead.  What do you need to be able to use this for your CI checks?

I don't have a strong opinion on the early-import point.  If there's something going wrong there, it's a bug to be fixed regardless of whether it's happening early or later.  If you'd like to work on moving imports earlier , that's great.  Please note that some of the `import` statements actually _execute_ functional code.  In those cases, the `import` is being used as a run-only-once construct.

Re the use of argparse and configparser:  I think using those is completely reasonable, so long as an interactive configuration capability is still there.  That capability doesn't need to look the same. It's valuable to be able to configure from inside the toolset without having to figure out how to construct a configuration file. Yes, I know that's not Linux-like or pythonic, but it's what I was told the existing DCC tools do, so it's a requirement.

**Comment by rm5248** [comment]
> Thanks for describing that in detail.
> 
> I've fixed a bug in the return codes, now in the HEAD of the `main` branch. I think the code now satisfies your first two points:
> 
> > Exit with a success code or failure code depending on if the test(s) passed or failed.
> 
> > Select which tests to run so it is not all-or-nothing
> 
> The selection can be done at several levels. You can run a single check with e.g. `./check_me10_init.py -r`, `./check_me20_verify.py -r`, etc (plus any necessary arguments to specify the connection, remote node ID, whether to run interactive tests, etc). Or you can run a family of checks with e.g. `./control_message.py -r` or run all families with `./control_master.py -r`

If I remember correctly when I was looking at it before the individual tests don't support command-line args.


> Top of my list on this right now is to fix the `control_*.py` checks that are still extra verbose when `-T 0` is specified. That flag should suppress normal, non-error output. AFAIK that's done properly in the individual `check_*.py` files, just not the `control_*.py` ones. But I could certainly work on something else instead. What do you need to be able to use this for your CI checks?

I didn't do any of this work because it would have touched a bunch of stuff, but using the python logger may make the most sense here, since it already does different levels of debug(instead of trying to re-invent the wheel).

For me, I think I can use what I currently have here in order to do some tests(it's currently rather low on my priority list).

> Re the use of argparse and configparser: I think using those is completely reasonable, so long as an interactive configuration capability is still there. That capability doesn't need to look the same. It's valuable to be able to configure from inside the toolset without having to figure out how to construct a configuration file. Yes, I know that's not Linux-like or pythonic, but it's what I was told the existing DCC tools do, so it's a requirement.

That sounds reasonable; the current interactive configuration is a little hard to parse mentally for me at least, but I don't have a better idea at the moment.

**Comment by bobjacobsen** [comment]
Thanks for the idea about using the `logging` package.  I've updated to using that and pushed to master. When you have a chance, could you try that and see if it works for your use case?

Working on an update to the configuration to make it more pythonic.  That's tricker than I originally thought, so it take some time.

Thanks for your efforts on this.
